Drop dependency on pkg/option in cilium-cni and cilium-docker

tklauser opened the pull request:
Move datapath mode, IPAM and ipvlan operation mode options out of `pkg/option` into their own special-purpose option packages (`pkg/datapath/option` and `pkg/ipam/option`, respectively). This allows to avoid dependencies on `pkg/option` and its transitive dependencies in several places. Namely, with a small internal API change in the last commit (2e6a52bfde474820fa3e0523c10f6a3110eca77c `datapath/linux: drop dependency on pkg/option` this reduces the binary size of `cilium-cni` and `cilium-docker` by ~1 MiB each (cf. #10056).

Please see individual commit messages for more details.

~Note: this should probably wait for #11083 being merged~ (was merged)

tklauser committed:
plugins/cilium-cni: avoid parsing CNI configuration twice in cmdAdd

There's already a call to types.LoadNetConf further up. This was
probably added by mistake in commit ff49865d477a ("plugins/cilium-cni:
disable CNI debug messages by default").

Signed-off-by: Tobias Klauser <tklauser@distanz.ch>

qmonnet reviewed (approved):
Nice work!

aanm commented on pkg/ipam/option/option.go:
```diff
@@ -0,0 +1,22 @@
+package option
```
Needs copyright

aanm commented on pkg/datapath/linux/node.go:
```diff
@@ -1148,7 +1148,7 @@ func (n *linuxNodeHandler) NodeValidateImplementation(nodeToValidate nodeTypes.N
 // handles the default route in the current namespace
 func NodeDeviceNameWithDefaultRoute() (string, error) {
-	link, err := route.NodeDeviceWithDefaultRoute()
+	link, err := route.NodeDeviceWithDefaultRoute(option.Config.EnableIPv4, option.Config.EnableIPv6)
```
Probably it would make sense to define a structure on each package that would specify its config? I'm thinking that if we add new options in the future that those won't make the function arguments increase as well.

tklauser replied:
Agree in general and I will also follow up with more PRs splitting config per package.

In this specific case I think we would still need to translate from `pkg/option.Config` to the to-be-introduced config for `pkg/datapath/linux/route` (since we cannot just move `EnableIPv{4,6}` to this new config, as it's used in a lot of other places). And we would make sure to not forget setting this config this at the right place. IMO, this would currently make things more complicated. And also, having exported functions depend on package-global variables IMO makes it harder to reason about these functions and test them.

Maybe we can address this once the need arises, i.e. once we would need to add more function arguments?

tklauser committed:
ipam: split IPAM mode option values into own package

These are used in several places where a full import of pkg/option (and
thus all its dependencies) can be avoided because only the
IPAM mode constants are used.

Together with the following commits, this helps to reduce the binary
size of cilium-cni and cilium-docker.

Signed-off-by: Tobias Klauser <tklauser@distanz.ch>

tklauser committed:
datapath: split datapath mode option values into own package

These are used in several places where a full import of pkg/option (and
thus all its dependencies) can be avoided because only the
DatapathModeVeth and DatapathModeIpvlan constants are used.

Together with the following commits, this helps to reduce the binary
size of cilium-cni and cilium-docker.

Signed-off-by: Tobias Klauser <tklauser@distanz.ch>

tklauser committed:
datapath/connector: move ipvlan operation mode options from pkg/option

The OperationModeL3 and OperationModeL3S consts are only used withing
pkg/datapath/connector and daemon/cmd. The former is also imported e.g.
in the cilium-cni and cilium-docker plugins. In order to avoid transitive
dependencies on pkg/option and its imports through
pkg/datapath/connector, move the option definitions there.

Signed-off-by: Tobias Klauser <tklauser@distanz.ch>

tklauser committed:
datapath/linux: drop dependency on pkg/option

With all the preceding commits, pkg/datapath/linux/route is now the only
package imported in the cilium-cni and cilium-docker plugins which pulls
in pkg/option. This can easily be avoided by changing
NodeDeviceWithDefaultRoute to not fetch the EnableIPv{4,6} options
itself from pkg/option, but by taking them as arguments.

Together with the preceding commits, this reduces the binary size of
cilium-cni and cilium-docker by ~1 MiB each:

before:

-rwxr-xr-x 1 tklauser tklauser 15364096 Mai  4 16:39 plugins/cilium-cni/cilium-cni
-rwxr-xr-x 1 tklauser tklauser 17399808 Mai  4 16:39 plugins/cilium-docker/cilium-docker

after:

-rwxr-xr-x 1 tklauser tklauser 13938688 Mai  4 17:00 plugins/cilium-cni/cilium-cni
-rwxr-xr-x 1 tklauser tklauser 16523264 Mai  4 17:00 plugins/cilium-docker/cilium-docker

Signed-off-by: Tobias Klauser <tklauser@distanz.ch>
